deprecate stacks, add targets (#582)
buildpacks/docs: content/docs/concepts/components/stack.md

]
+++

## What is a stack and Why is it deprecated?

### Deprecation
A stack restricts buildpacks to run on top of a set of related **build images** (where your app is built)
and **run images** (where your app runs).  Practical experience of the buildpacks community has not found that the stack concept adds any value therefore [the community decided](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md)
to use buildpacks with build and run images without the additional concept formerly known as _stack_. 

Stacks will continue to be supported with full backwards compatibility.

### What's Next?
[Targets](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md#example-buildpacktoml-targets-table) replace stacks.

> Targets allow buildpack authors to directly specify details such as OS and architecture directly without the intermediate object known as a "stack." 

Reviewer: ```suggestion
Targets allow buildpack authors to directly specify details such as OS and architecture directly without the intermediate object known as a "stack".
```

Targets are available starting with the 0.12 Platform API but should be safely ignored by older platforms.

#### How long can I keep using stacks?
We do not yet have a defined end-date. Historically this project allows deprecated features to remain for 6-24+ months.

#### What is the current recommended best practice?
We currently recommend that buildpack authors use both a stack and a target.
We anticipate it will take some time before all platforms catch up with this change, but by using both your buildpacks will work with all platforms past, present, and future.

In order to ease this process for those using the io.buildpacks.stacks.bionic, lifecycle will translate any section that sets this as on of the stacks:

```
[[stacks]]

> id = "io.buildpacks.stacks.bionic

Reviewer 2: ```suggestion
id = "io.buildpacks.stacks.bionic"
```

]
+++

## What is a stack and Why is it deprecated?

### Deprecation

> Stacks are great - long live the stack! 
> However, a stack is basically an extra layer on top of the **build image** (where your app is built)
> and the **run image** (where your app runs) and [the community decided](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md)
> to use those two concepts directly without the additional layer formerly known as _stack_. 

Reviewer: I wonder if it makes sense to compare a stack to a layer?  End-users might expect to find that layer when they run `crane` or some similar tool.

```suggestion
A stack restricts buildpacks to run on top of a set of related **build images** (where your app is built)
and **run images** (where your app runs).  Practical experience of the buildpacks community has not found that the stack concept adds any value therefore [the community decided](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md)
to use buildpacks with build and run images without the additional concept formerly known as _stack_. 
```

Stacks will continue to be supported with full backwards compatibility.

### What's Next?
[Targets](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md#example-buildpacktoml-targets-table) replace stacks.
Targets allow buildpack authors to directly specify details such as OS and architecture directly without the intermediate object known as a "stack." 
Targets are available starting with the 0.12 Platform API but should be safely ignored by older platforms.

#### Wait - are you cutting me off?

> My friend, never fear. We have a long horizon and a heartfelt desire to ease your transition. We will provide backwards compatibility with stacks for so long that we haven't even decided yet when we're going to turn them off.

Reviewer: I like the more personal style that you use.  However we're much more straight-laced in the rest of the docs.  Should we prefer the (more boring) existing style?

```suggestion
My friend, never fear. We have a long horizon and a heartfelt desire to ease your transition. We will provide backwards compatibility with stacks for so long that we haven't even decided yet when we're going to turn them off.
```

Author: oh yes, almost certainly.
